### PR TITLE
Fully dockerized flow skeleton

### PR DESCRIPTION
**Pull request by abaptiste**
This change adds a skeleton for the fully dockerized benchmark flow.

This benchmark method uses docker images for  Envoy, the NightHawk benchmark framework, and the NightHawk binaries.  

Users specify a location where external tests are mapped into the benchmark container and executed.

cc: @mum4k, @oschaaf, @htuch 

Signed-off-by: Alvin Baptiste <alvinsb@gmail.com>

### CODE REVIEW

File: salvo/src/lib/benchmark/base_benchmark.py
Comment (mum4k) [inline salvo/src/lib/benchmark/base_benchmark.py:17]: A question on privacy - are all of these functions and method meant to be public? (will they be used from outside of this package?

If some of them aren't we should prefer to make them private by prepending an underscore to their name. Private methods make it easier to change code around in the future.

https://github.com/google/styleguide/blob/gh-pages/pyguide.md#3162-naming-conventions

Comment (mum4k) [inline salvo/src/lib/benchmark/base_benchmark.py:17]: Assuming the test_dir is meant to be a string, can we try to also keep its default value within the same object type? Even though Python isn't type safe, it is good to follow the generic type safety rules to avoid unexpected programming errors, like when someone tries to call a string method on a NoneType.

```py
def get_docker_volumes(output_dir, test_dir=''):
```

File: salvo/src/lib/benchmark/base_benchmark.py
Comment (mum4k) [inline salvo/src/lib/benchmark/base_benchmark.py:1]: (optional, food for thought) How would you feel about using pytype annotations in code in this repository? They tend to help a lot by being able to statically discover most common programming errors in Python.

Let me know if this is something you would be willing to explore, I can provide some hints.

Some background:
https://github.com/google/styleguide/blob/gh-pages/pyguide.md#221-type-annotated-code

File: salvo/src/lib/benchmark/fully_dockerized_benchmark.py
Comment (mum4k) [inline salvo/src/lib/benchmark/fully_dockerized_benchmark.py:17]: What arguments are we expecting to be passing here and can we name them instead of using `kwargs`? Listing the arguments specifically would help us to think about code structure and would make the code more readable.

File: salvo/src/lib/benchmark/fully_dockerized_benchmark.py
Comment (mum4k) [inline salvo/src/lib/benchmark/fully_dockerized_benchmark.py:20]: Since this is a public method, are we expecting the callers to call it specifically? Is that the right approach or should it be a private method that we always call from say `execute_benchmark`?

Other way of asking the same is - are there any situations where the caller would want to execute without validating? If we call it internally we may get better safety (impossible to execute without validation) and smaller API surface.

**Review by mum4k — CHANGES_REQUESTED** [review]
Thank you for splitting the API into its own PR and keeping the scope limited.

Added a few comments and some questions. As discussed previously - please tag me on the PR specifically when all comments will be responded to and you want me to take another look.

File: salvo/api/env.proto
Comment (abaptiste) [reply salvo/api/env.proto:12]: This makes sense.  I will update

File: salvo/api/source.proto
Comment (abaptiste) [reply salvo/api/source.proto:10]: ack

File: salvo/src/lib/BUILD
Comment (abaptiste) [reply salvo/src/lib/BUILD:10]: Let me try this.  I tried some variations based on the docs and got errors about circular dependencies.

File: salvo/src/lib/benchmark/BUILD
Comment (abaptiste) [reply salvo/src/lib/benchmark/BUILD:15]: ack

File: salvo/src/lib/benchmark/base_benchmark.py
Comment (abaptiste) [reply salvo/src/lib/benchmark/base_benchmark.py:12]: certainly

File: salvo/src/lib/benchmark/base_benchmark.py
Comment (abaptiste) [reply salvo/src/lib/benchmark/base_benchmark.py:17]: Yep, I can take a look and make the functions private where applicable

Comment (abaptiste) [reply salvo/src/lib/benchmark/base_benchmark.py:17]: sure

File: salvo/src/lib/benchmark/base_benchmark.py
Comment (abaptiste) [reply salvo/src/lib/benchmark/base_benchmark.py:1]: I can investigate and add this later once I address the other comments.

File: salvo/src/lib/benchmark/fully_dockerized_benchmark.py
Comment (abaptiste) [reply salvo/src/lib/benchmark/fully_dockerized_benchmark.py:17]: Sure.  For now it's just one argument.  I will update.   There are other places where `kwargs` is used like this so I'll take a look at those also.

File: salvo/src/lib/benchmark/fully_dockerized_benchmark.py
Comment (abaptiste) [reply salvo/src/lib/benchmark/fully_dockerized_benchmark.py:20]: Right, this is true.  We shouldn't rely on the user of the class to call validate.  I'll make this private and update the PR

**Comment by abaptiste** [comment]
@mum4k This is ready for another pass.  Thanks.

I am still looking into the Type Hints and will chat with you about that in slack.

File: salvo/src/lib/benchmark/base_benchmark.py
Comment (mum4k) [reply salvo/src/lib/benchmark/base_benchmark.py:1]: (optional) We certainly don't have to add pytype annotations in this PR (and we aren't going to block this PR on this comment).

I did find the pytype annotations very helpful in preventing bugs and shortening the debug cycle on Python code, so if you are interested and inclined towards using them here I will be happy to work with you and help you figure out how to configure the build system to make them work.

Let's address this in a slack conversation as you have suggested.

File: salvo/src/lib/benchmark/fully_dockerized_benchmark.py
Comment (mum4k) [inline salvo/src/lib/benchmark/fully_dockerized_benchmark.py:17]: Looks like the comment about `kwargs` was only partially addressed. We do have exact arguments now, but we kept the `kwargs` also.

Does this mean we don't know the complete list of arguments here? And if that is the case, can we discuss this and figure out an architecture where the full list of arguments passed through the layers is known?

If we do know the complete list of arguments, can we remove the `kwargs`?

As you have pointed out - this applies to a few locations in this PR.

Comment (mum4k) [inline salvo/src/lib/benchmark/fully_dockerized_benchmark.py:17]: Note, an initializer of a public class is also considered public and should have a docstring. When updating the other docstrings to the format recommended by the style guide, let's also add a docstring here and include an `Args:` section where we describe the arguments the initializer takes.

**Review by mum4k — CHANGES_REQUESTED** [review]
Thanks for iterating, looks like a couple of the review comments need a bit more attention, or need to be applied in more locations. I have left comments in location where this applies.

File: salvo/src/lib/benchmark/fully_dockerized_benchmark.py
Comment (abaptiste) [reply salvo/src/lib/benchmark/fully_dockerized_benchmark.py:17]: I'll remove it here since it's unused.  The other places pass kwargs to things like `subprocess.check_output` and the method to run a docker image.  In those cases kwargs seems much more compact than having a long list of parameters and values.

**Comment by abaptiste** [comment]
@mum4k Addressed your comments.  Please review when you have a chance.  Thanks!

File: salvo/src/lib/job_control_loader.py
Comment (mum4k) [inline salvo/src/lib/job_control_loader.py:25]: When describing a complex container, we should be more explicit about its structure and the inner object types to help the reader / user.

Example for this dictionary: {'John Doe': 345234}

Returns:
  A dict of strings to integers, the keys are names of people and the
  values are their phone numbers. Example:
    {'John Doe': 345234}

Including an example section is optional and I would only include it for more complex dictionaries.

Note - this is another requirement that goes away if we choose to use the pytype annotations.

This comment applies to a few locations.

**Review by mum4k — APPROVED** [review]
Thank you @abaptiste, this looks good. Including the last set of comments.

@htuch please review and merge once the remaining comments are resolved.

File: salvo/src/lib/benchmark/base_benchmark.py
Comment (abaptiste) [reply salvo/src/lib/benchmark/base_benchmark.py:21]: updated

File: salvo/src/lib/constants.py
Comment (abaptiste) [reply salvo/src/lib/constants.py:5]: Updated

File: salvo/src/lib/job_control_loader.py
Comment (abaptiste) [reply salvo/src/lib/job_control_loader.py:25]: Can we revisit this point in a later PR. I have pytypes working locally and can add annotations to this module:

```
$ grep ^"def _load_json_doc" src/lib/job_control_loader.py
def _load_json_doc(filename: str) -> JobControl:
$ pytype src -P bazel-bin:.
Computing dependencies
Analyzing 25 sources with 6 local dependencies
ninja: Entering directory `/home/ubuntu/code/salvo/.pytype'
[31/31] check src.lib.service.test_s3_upload_event
Leaving directory '/home/ubuntu/code/salvo/.pytype'
Success: no errors found

```

Comment (abaptiste) [reply salvo/src/lib/job_control_loader.py]: ptypes are coming.  I was able to get this going locally.

Comment (mum4k) [reply salvo/src/lib/job_control_loader.py]: Great news! Note that even with pytype, we should still say the meaning of the keys and values. Pytype removes the need to specifically mention the object types.

Something like:

```
Returns:
  A dictionary, keys are the file names and values are the file content.
```

**Comment by abaptiste** [comment]
@htuch This is ready for your review.  Thanks!

**Comment by htuch** [comment]
@envoyproxy/senior-maintainers agreed to make @mum4k a maintainer of this repo (based on Nighthawk history), so I think we can merge once he is satisfied. Thanks.

**Comment by mum4k** [comment]
Thank you @htuch, this will decrease the number of review rounds on each PR and help us maintain a good velocity while developing Salvo. We will certainly keep you in the loop for any major architecture decisions or direction changes.

@abaptiste if you can, please address the remaining comment about documenting the meaning of keys and values of some of the dictionaries. We can also do that in a later PR if you want. I will give this PR one more glance and merge later tonight.

File: salvo/src/lib/benchmark/base_benchmark.py
Comment (abaptiste) [reply salvo/src/lib/benchmark/base_benchmark.py:20]: I would prefer to go with pytype annotations.  I'm not certain whether the action here is to add the object types to the documentation in addition to the annotations.

I do have the annotations working and can include those in the next PR.

**Review by mum4k — APPROVED** [review]

**Pull request merged**
